lasagne/regularization.py

```diff
@@ -1,15 +1,47 @@
 import theano.tensor as T
+from .layers import Layer, get_all_params
 
-from . import layers
 
+def l1(param):
+    """ L1 weight decay penalty """
+    return T.sum(abs(param))
 
-def l2(layer, include_biases=False):
-    if include_biases:
-        all_params = layers.get_all_params(layer)
-    else:
-        all_params = layers.get_all_non_bias_params(layer)
 
-    return sum(T.sum(p**2) for p in all_params)
+def l2(param):
+    """ L2 weight decay penalty """
+    return T.sum(param**2)
 
 
-# TODO: sparsity regularization
+def apply_penalty(params, penalty, **kwargs):
+    try:
+        if len(params) == 0:
+            return 0
+        else:
+            return sum(penalty(p, **kwargs) for p in params)
+    except TypeError:
+        return penalty(params, **kwargs)
+
+
+def regularize_layer_params(layer, penalty,
+                            tags={'regularizable': True}, **kwargs):
+    layers = [layer, ] if isinstance(layer, Layer) else layer
+    all_params = []
+
+    for layer in layers:
+        all_params += layer.get_params(**tags)
```

Comment:
> Are we anticipating many penalty functions with keyword arguments to control their behaviour? 

One is enough. But are we anticipating many people will use the shortcut with a different set of tags?

> It's a bit unfortunate that the tag interface is different for this one function though. I wonder if there's a way around it.

One way would be using a dictionary in `get_params()` and `get_all_params()` as well. Another way would be not to allow keyword arguments being passed on to the penalty function, such that customizable penalty functions must be realized as callable classes (similar to `LeakyRectify` in `lasagne.nonlinearities`).

> Also, if we use `**tags`, how would the user specify the absence of the `regularizable` tag?

Right, it would limit usage to `regularizable=True` and `regularizable=False`. To collect all parameters independent of the `regularizable` tag, we'd need to invent an extra case such as `regularizable=None`. That probably wouldn't make things any better than having `tags` be a dictionary, unless we deem the `None` case to be generally useful and change `get_params()` accordingly as well.